File: demo/src/java/eval/KoKoList.java
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.ast;
import java.util.*;
import java.io.*;

public class KoKoList extends ArrayList<KoKoAst> implements KoKoAst{
    public KoKoList(List<KoKoAst> list){
        super(list);
    }

    public KoKoList(){
        super();
    }

    public void genCode(PrintStream out){
        if(this.size() == 0) return;
        this.get(0).genCode(out);
        this.stream()
            .skip(1)
            .forEach( t -> {
                            out.print(", ");
                            t.genCode(out);
            });
    }

    public KoKoValue eval(KoKoContext ctx){
        KoKoListValue res = new KoKoListValue();
        for(int i = 0; i < this.size(); i++){
            res.add(this.get(i).eval(ctx));
        }
        return res;
    }

    public KoKoValue eval(){
        return eval(new KoKoContext());
    }
} 

File: KoKoslan/src/kokoslan/kotlin/eval/KoKoList.kt
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.kotlin.eval

import java.util.*
import java.io.*
import kokoslan.kotlin.ast.*


class KoKoList(var list : List<KoKoAst>? = null, private var kokoNativeList: Boolean = false) : ArrayList<KoKoAst>(list), KoKoAst{
    
    override fun genCode(Out : PrintStream){
        if(this.size == 0) return
        if(kokoNativeList)
            Out.print("[ ")
        this[0].genCode(Out)
        this.drop(1)
            .forEach{Out.print(", "); it.genCode(Out);}
        if(kokoNativeList)
            Out.print("] ")
    }

    override fun eval(ctx : KoKoContext) : KoKoValue? {
        val res = KoKoListValue()
        (0 until this.size).mapTo(res) { this[it].eval(ctx) }
        return res
    }

    fun eval() : KoKoValue? {
        return eval(KoKoContext())
    }
    
} 